modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/CereProcEngine.java

```diff
@@ -202,4 +213,133 @@ public TTSResource allocateThreadResources() throws SynthesisException, Interrup
+		XslTransformCompiler xslCompiler = new XslTransformCompiler(conf);
+		CompiledStylesheet styleSheet;
+		styleSheet = this.mStyleSheets.get(conf);
+		if (styleSheet == null) {
+			try {
+				styleSheet = xslCompiler.compileStylesheet(CereProcEngine.class.getResourceAsStream("/transform-ssml.xsl"));
+				mStyleSheets.put(conf, styleSheet);
+			} catch (SaxonApiException e) {
+				logger.error(e.toString());
+				throw new SynthesisException(e);
+			}
+		}
+		return styleSheet;
+	}
+
+	private void performSubstitutionRules(XMLStreamReader reader, XMLStreamWriter writer, String lang) throws XMLStreamException {
+		int i ,len;
+		while(reader.hasNext()) {
+			reader.next();
+			int et = reader.getEventType();
+			switch (et) {
+				case XMLEvent.START_ELEMENT:
+					final String localName = reader.getLocalName();
+					final String namespaceURI = reader.getNamespaceURI();
+					if (namespaceURI != null && namespaceURI.length() > 0) {
+						final String prefix = reader.getPrefix();
+						if (prefix != null)
+							writer.writeStartElement(prefix, localName, namespaceURI);
+						else
+							writer.writeStartElement(namespaceURI, localName);
+					} else {
+						writer.writeStartElement(localName);
+					}
+
+					for ( i = 0, len = reader.getNamespaceCount(); i < len; i++) {
+						writer.writeNamespace(reader.getNamespacePrefix(i), reader.getNamespaceURI(i));
+					}
+
+					for ( i = 0, len = reader.getAttributeCount(); i < len; i++) {
+						String attUri = reader.getAttributeNamespace(i);
+						if (attUri != null)
+							writer.writeAttribute(attUri, reader.getAttributeLocalName(i), reader.getAttributeValue(i));
+						else
+							writer.writeAttribute(reader.getAttributeLocalName(i), reader.getAttributeValue(i));
+					}
+					break;
+				case XMLEvent.END_ELEMENT:
+					writer.writeEndElement();
+					break;
+				case XMLEvent.SPACE:
+				case XMLEvent.CHARACTERS:
+					CereprocTTSUtil utils = this.mTtsUtils.get(lang);
+					writer.writeCharacters(utils.applyAll(reader.getText()));
+					break;
+				case XMLEvent.PROCESSING_INSTRUCTION:
+					writer.writeProcessingInstruction(reader.getPITarget(), reader.getPIData());
+					break;
+				case XMLEvent.CDATA:
+					writer.writeCData(reader.getText());
+					break;
+				case XMLEvent.COMMENT:
+					writer.writeComment(reader.getText());
+					break;
+				case XMLEvent.ENTITY_REFERENCE:
+					writer.writeEntityRef(reader.getLocalName());
+					break;
+				case XMLEvent.START_DOCUMENT:
+					String encoding = reader.getCharacterEncodingScheme();
+					String version = reader.getVersion();
+
+					if (encoding != null && version != null)
+						writer.writeStartDocument(encoding, version);
+					else if (version != null)
+						writer.writeStartDocument(reader.getVersion());
+					break;
+				case XMLEvent.END_DOCUMENT:
+					writer.writeEndDocument();
+					break;
+				case XMLEvent.DTD:
+					writer.writeDTD(reader.getText());
+					break;
+			}
```

Comment:
This code can be simplified a lot by using the utility function [`org.daisy.common.stax.XMLStreamWriterHelper.writeEvent`](http://daisy.github.io/pipeline/api/org/daisy/common/stax/XMLStreamWriterHelper.html#writeEvent-javax.xml.stream.XMLStreamWriter-javax.xml.stream.XMLStreamReader-). Sorry, I should have pointed you to this beforehand.

Reply:
Fixed!